packages/exo/src/exo-tools.js

```diff
@@ -6,59 +6,150 @@ import {
 /**
  * @template {Record<PropertyKey, MethodGuard>} [T=Record<PropertyKey, MethodGuard>]
  * @typedef {import('@endo/patterns').InterfaceGuard<T>} InterfaceGuard
  */
-/** @typedef {import('@endo/patterns').InterfaceGuardKit} InterfaceGuardKit */
 
 const { quote: q, Fail } = assert;
 const { apply, ownKeys } = Reflect;
 const { defineProperties, fromEntries } = Object;
 
+/**
+ * A method guard, for inclusion in an interface guard, that does not
+ * enforce any constraints of incoming arguments or return results.
+ */
+const RawMethodGuard = M.call().rest(M.raw()).returns(M.raw());
+
+const REDACTED_RAW_ARG = '<redacted raw arg>';
+
 /**
  * A method guard, for inclusion in an interface guard, that enforces only that
  * all arguments are passable and that the result is passable. (In far classes,
- * "any" means any *passable*.) This is the least possible enforcement for a
- * method guard, and is implied by all other method guards.
+ * "any" means any *passable*.) This is the least possible non-raw
+ * enforcement for a method guard, and is implied by all other
+ * non-raw method guards.
+ */
+const PassableMethodGuard = M.call().rest(M.any()).returns(M.any());
+
+/**
+ * @typedef {object} RedactConfig
+ * @property {number} declaredLen
+ * @property {boolean} hasRestArgGuard
+ * @property {Pattern} paramsPattern
+ * @property {boolean} rawRestArgs
+ * @property {number[]} redactedIndices
  */
-const MinMethodGuard = M.call().rest(M.any()).returns(M.any());
 
 /**
  * @param {Passable[]} syncArgs
- * @param {MethodGuardPayload} methodGuardPayload
+ * @param {RedactConfig} redactConfig
  * @param {string} [label]
  * @returns {Passable[]} Returns the args that should be passed to the
  * raw method
  */
-const defendSyncArgs = (syncArgs, methodGuardPayload, label = undefined) => {
-  const { argGuards, optionalArgGuards, restArgGuard } = methodGuardPayload;
-  const paramsPattern = M.splitArray(
-    argGuards,
-    optionalArgGuards,
-    restArgGuard,
-  );
-  mustMatch(harden(syncArgs), paramsPattern, label);
-  if (restArgGuard !== undefined) {
+const defendSyncArgs = (syncArgs, redactConfig, label = undefined) => {
+  const {
+    declaredLen,
+    hasRestArgGuard,
+    paramsPattern,
+    rawRestArgs,
+    redactedIndices,
+  } = redactConfig;
+
+  // Only manipulate the original args if our params pattern didn't contain any
+  // raw values.
+  let redactedArgs = syncArgs;
+  if (rawRestArgs) {
+    // Don't harden the rest args.
+    redactedArgs = syncArgs.slice(0, Math.min(syncArgs.length, declaredLen));
+  } else if (redactedIndices.length > 0) {
+    // Copy the arguments array, avoiding hardening the redacted ones (which are
+    // trivially matched).
+    redactedArgs = [...syncArgs];
+  }
+
+  for (const i of redactedIndices) {
+    if (i >= redactedArgs.length) {
+      break;
+    }
+    redactedArgs[i] = REDACTED_RAW_ARG;
+  }
+
+  mustMatch(harden(redactedArgs), paramsPattern, label);
```

Comment:
The `redactedArgs` name seems a little misleading, implying e.g. "those arguments that have been redacted". Maybe instead `matchableArgs`?

Also, I question the value of completely ignoring raw rest args here—it's probably better to preserve their presence:
```js
  // Use syncArgs if possible, but copy it when necessary to implement redactions.
  let matchableArgs = syncArgs;
  if (rawRestArgs && syncArgs.length > declaredLen) {
    const restLen = syncArgs.length - declaredLen;
    const redactedRest = Array(restLen).fill(REDACTED_RAW_ARG);
    matchableArgs = [...syncArgs.slice(0, declaredLen), ...redactedRest];
  } else if (redactedIndices.length > 0 && redactedIndices[0] < syncArgs.length) {
    matchableArgs = [...syncArgs];
  }

  for (const i of redactedIndices) {
    if (i >= matchableArgs.length) {
      break;
    }
    matchableArgs[i] = REDACTED_RAW_ARG;
  }

  mustMatch(harden(redactedArgs), paramsPattern, label);
```